### DISCUSSION TITLE
[SUGGESTION] SL Web Tools implementation as Web Flasher for flashing Silicon Labs EFR32 firmware via a web browser

### DISCUSSION THREAD

**Discussion by Hedda** [tube0013/tube_gateways#135] [Ideas]
Home Assistant founders (and Nabu Casa) have recently among other things announced: "[SL Web Tools](https://github.com/NabuCasa/sl-web-tools)" project (short for "Silicon Labs Web Tools") is available as open source, so wondering if you would consider adding that to [tubeszb.com](https://tubeszb.com) website as a tool for updating/flashing [EFR32 MGM21 Ethernet/USB Serial Coordinator](https://tubeszb.com/product/efr32-mgm21-ethernetusb-serial-coordinator/)

https://www.home-assistant.io/blog/2023/02/08/state-of-matter-and-thread/#silabs-multi-flasher--sl-web-tools

(Note that this is only just one section from https://www.home-assistant.io/blog/2023/02/08/state-of-matter-and-thread/ )

SL Web Tools allows users to manage the firmware on Zigbee/Thread sticks directly from a web browser, with no terminals or compilation necessary. This allows product creators like yourself to offer easy firmware updates to their users and make experimental firmware updates available to test bug fixes quickly.

SL Web Tools runs in the browser using [Pyodide](https://pyodide.org/en/stable/) and is currently powered by another new open-source Python project from them, “[silabs-universal-flasher](https://github.com/NabuCasa/universal-silabs-flasher)” for Silicon Labs EFR32 based SoCs like the EFR32MG21 SoC.

https://github.com/NabuCasa/sl-web-tools

(and currently powered by https://github.com/NabuCasa/universal-silabs-flasher for Silicon Labs based chips)

https://www.youtube.com/watch?v=-88K23e8XYw&t=2s&ab_channel=HomeAssistant

To see it in action, a demo of SL Web Tools for Silicon Labs chips has been added to the Home Assistant SkyConnect documentation.

https://skyconnect.home-assistant.io/firmware-update/ redirected from https://skyconnect.home-assistant.io/ 

![image](https://user-images.githubusercontent.com/6320001/220318372-651a0e8b-081a-4ce8-a8bc-38da13a520cd.png)

Also, if wanted to take it to the next level then maybe forking that and make a project called "TI Web Tools" (short for "Texas Instruments Web Tools") or similar that is based on that in combination with [cc2538-bsl](https://github.com/JelmerT/cc2538-bsl) code (and/or [ZigStarGW-MT](https://github.com/xyzroe/ZigStarGW-MT)) for the TI flashing part(?). That is, an idea to extend this would be to replace the [silabs-universal-flasher](https://github.com/NabuCasa/universal-silabs-flasher) code part with [cc2538-bsl](https://github.com/JelmerT/cc2538-bsl) and/or [ZigStarGW-MT](https://github.com/xyzroe/ZigStarGW-MT) code in order to allow flashing of Texas Instruments based adapters instead if you think that would be possible?

Note! [SL Web Tools](https://github.com/NabuCasa/sl-web-tools) is in turn a fork of their previous "[ESP Web Tools](https://esphome.github.io/esp-web-tools/)" project which allow DIY firmware flashing of ESP32/ESP8266 SoC projecst to offer an easy web installer on their homepage, allowing users to get started directly from their browser. It has been adopted by all the great projects, including [Tasmota](https://tasmota.github.io/install/), [WLED](https://install.wled.me/), and [ESPresense](https://espresense.com/firmware). Ever since the introduction of ESP Web Tools, they expanded this to more devices, and now they also introduced [SL Web Tools](https://github.com/NabuCasa/sl-web-tools) for SiLabs based chips.

PS: For reference; the firmware build system for Silicon Labs was made by Home Assistant developer/engineer and Nabu Casa employee [Stefan Agner (a.k.a. agners](https://github.com/agners), who also developed the EFR32MG21-based [Home Assistant SkyConnect USB stick/dongle](https://www.home-assistant.io/skyconnect) as well as the EFR32MG21-based radio module integrated into the Home Assistant Yellow appliance for Rasberry Pi 4 Compute Module). And the "universal-silabs-flasher" firmware tool was made by [puddly](https://github.com/puddly) who is also one of the lead developers of [zigpy ](https://github.com/zigpy/zigpy)and the [radio libraries for zigpy](https://github.com/zigpy) that is used in Home Assistant's [ZHA integration](https://www.home-assistant.io/integrations/zha) (among other projects).

**Comment by tube0013** [comment]
Web flasher only works with usb connected devices. It cannot currently access local network devices.

This may or may not happen as it is basically a browser security feature.

**Reply by Hedda** [reply]
Sorry I meant your for your "EFR32 MGM21 Ethernet/USB Serial Coordinator" (flashing over USB/serial) -> https://tubeszb.com/product/efr32-mgm21-ethernetusb-serial-coordinator/

**Reply by tube0013** [reply]
Also my understanding is there is only one up stream bug preventing ti flashing.

**Reply by Hedda** [reply]
But is your EFR32 MGM21 Ethernet/USB Serial Coordinator not a Silicon Labs EFR32MGM21 based Zigbee Coordinator that can be connected over Ethernet or USB (e.i. serial-over-USB?)? https://tubeszb.com/product/efr32-mgm21-ethernetusb-serial-coordinator/

**Reply by tube0013** [reply]
it should work with the one provided by Nabu Casa, However I need to check with Puddly as to why only USB serial CP210x is detected and not CH340 based which is what
![image](https://user-images.githubusercontent.com/11084412/220380496-ce0f2e30-1bf3-4b76-bfed-568320a600f6.png)
 I use.

**Reply by MattWestb** [reply]
For my is the same SC is detected but not other EZSP devices that is having CH340  USB-chips.
I was thinking  the web-flasher  is locked on SC USB-ID.

**Reply by tube0013** [reply]
looks to be the case: https://github.com/NabuCasa/sl-web-tools/blob/dc5d5c239d330e03add1511971280e4180bebf94/assets/manifests/skyconnect.json#L5

**Reply by MattWestb** [reply]
Then doing one fork and adding little more IDs or asking Puddly doinng one "open" version.

**Reply by tube0013** [reply]
it is open already, it's just a matter of implementing it on your own site. I'm not going to spend the time unless Network flash works. I've discussed it with Puddly he had some thoughts but it's unknown if it will work due to reasons I said before.
